zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationCheckBehavior.java

```diff
@@ -50,36 +57,61 @@ public AuthorizationCheckBehavior(
    * contains the data required to do the check.
    *
    * @param request the authorization request to check authorization for. This contains the command,
-   *     the resource type, the permission type and a set of resource identifiers
-   * @return true if the entity is authorized, false otherwise
+   *     the resource type, the permission type, a set of resource identifiers and the tenant id
+   * @return a {@link Either} containing a {@link RejectionType} if the user is not authorized or
+   *     {@link Void} if the user is authorized
    */
-  public boolean isAuthorized(final AuthorizationRequest request) {
+  public Either<RejectionType, Void> isAuthorized(final AuthorizationRequest request) {
     if (!engineConfig.isEnableAuthorization()) {
-      return true;
+      return Either.right(null);
     }
 
     if (!request.getCommand().hasRequestMetadata()) {
       // The command is written by Zeebe internally. Internal Zeebe commands are always authorized
-      return true;
+      return Either.right(null);
     }
 
     Set<String> authorizedResourceIdentifiers = new HashSet<>();
 
     final var userKey = getUserKey(request);
     if (userKey.isPresent()) {
+      final var userOptional = userState.getUser(userKey.get());
+      // verify if the user is authorized for the tenant
+      if (!isUserAuthorizedForTenant(request, userOptional)) {
+        return Either.left(RejectionType.NOT_FOUND);
+      }
+
       authorizedResourceIdentifiers =
           getUserAuthorizedResourceIdentifiers(
-              userKey.get(), request.getResourceType(), request.getPermissionType());
+              userOptional, request.getResourceType(), request.getPermissionType());
     }
 
     // Check if authorizations contain a resource identifier that matches the required resource
     // identifiers
-    return hasRequiredPermission(request.getResourceIds(), authorizedResourceIdentifiers);
+    if (hasRequiredPermission(request.getResourceIds(), authorizedResourceIdentifiers)) {
+      return Either.right(null);
+    } else {
+      return Either.left(RejectionType.UNAUTHORIZED);
+    }
   }
 
   private static Optional<Long> getUserKey(final AuthorizationRequest request) {
-    return Optional.ofNullable(
-        (Long) request.getCommand().getAuthorizations().get(Authorization.AUTHORIZED_USER_KEY));
+    return getUserKey(request.getCommand());
+  }
+
+  private boolean isUserAuthorizedForTenant(
+      final AuthorizationRequest request, final Optional<PersistedUser> userOptional) {
```

Comment:
🔧 It's a bad practice to pass optionals as parameters. When you do this you will have always have to introduce some conditional flow in the method in case the optional is not set. Since you now pass it as a parameter to two methods it means both these methods have to deal with this potentially empty optional.

Instead what we should do is when we get this user check if we actually found it. If not we can just reject it with an unauhorized straight away. In this method and the other method that received this optional we can change it to receive the user. because we know we'll have found it at that point in time.

Reply:
I did that because the behaviour is not the same for the two methods, in the tenant check we return false (that means that the user is unauthorized basically). In the other one we return an empty set and then we continue with verify the permission. 
I can't reject the command if the user is not found, otherwise I'll change the behaviour of the method

Reply:
The empty set achieves the same thing. It means we have no authorized resource identifiers. That means when we do the comparison it will never have the required permissions and will always be unauthroized. Using the empty set was just a convenient way to deal with this at that time.

```java
  private boolean hasRequiredPermission(
      final Set<String> requiredResourceIdentifiers,
      final Set<String> authorizedResourceIdentifiers) {
    return authorizedResourceIdentifiers.stream().anyMatch(requiredResourceIdentifiers::contains);
  }
```

You can safely say that if there is no user, it is not authorized.

Reply:
I'll change it